src/allreduce_robust.h

```diff
@@ -177,64 +199,91 @@ class AllreduceRobust : public AllreduceBase {
     // check point Ack, we use a two phase message in check point,
     // this is the second phase of check pointing
     static const int kCheckAck = 4;
+    // some node want to load cache
+    static const int kLoadCache = 8;
     // there are difference sequence number the nodes proposed
     // this means we want to do recover execution of the lower sequence
     // action instead of normal execution
-    static const int kDiffSeq = 8;
+    static const int kDiffSeq = 16;
     // constructor
     ActionSummary(void) {}
     // constructor of action
-    explicit ActionSummary(int flag, int minseqno = kSpecialOp) {
-      seqcode = (minseqno << 4) | flag;
+    explicit ActionSummary(int action_flag, int role_diff_flag = 0,
+      int minseqno = kSpecialOp, int maxseqno = kSpecialOp) {
```

Comment:
why we need to regulate max/min seq no

Reply:
I don't think there is something like int overflow, 

if seqno is increment by 1 for each operation, in fast histo,  for depthwise growing, (for iterations after the first one) we only have 1 sync for histogram, for loss-guided, we have at most number of leaf nodes sync; for approx, the number of sync is tree's height, 

why it is overflowing?

Reply:
>why we need to regulate max seq no 
maxseqno is keeping cache_seq and seperate itself from allreduce seqno, what it does is to get number of cache entries in healthy and decide if backfill recovered host is needed.

>why it is overflowing?
First what overflow is minseqno https://github.com/dmlc/xgboost/issues/4250#issuecomment-510745434. what minseqno does is to keep minimal seqno within all nodes, with one exception. 

kSpecialOp which is very large number  https://github.com/dmlc/rabit/blob/master/src/allreduce_robust.h#L165
when it construct, it bitwise move 4 digits to leave space for four masks, latter move 5 digits as we tried to add one more mask. 

https://github.com/dmlc/rabit/blob/master/src/allreduce_robust.h#L188
This is likely cause overflow as singed int type (1 << 26) << 5 > 2^31-1

Reply:
why the relationship is defined as `min` and `max`? you just distinguish operations from different sources

same for SeqType, why it is called OR/AND ?

Reply:
It was documented in detail in design doc. Simply put, cache behavior is different from normal seq no based catch up.
https://github.com/dmlc/xgboost/files/3482936/Rabit.bootstrap.operations.support.pdf